testslide/strict_mock.py

```diff
@@ -581,17 +581,21 @@ def _validate_and_wrap_mock_value(self, name, value):
                             name,
                             self.__dict__["_type_validation"],
                         )
+
                         if inspect.iscoroutinefunction(template_value):
 
                             async def awaitable_return_validation_wrapper(
                                 *args, **kwargs
                             ):
-                                return_value = signature_validation_wrapper(
+                                result_awaitable = signature_validation_wrapper(
                                     *args, **kwargs
                                 )
-                                if not inspect.isawaitable(return_value):
+                                if not inspect.isawaitable(result_awaitable):
                                     raise NonAwaitableReturn(self, name)
-                                return await return_value
+                                return_value = await result_awaitable
+                                if self.__dict__["_type_validation"]:
+                                    testslide.lib._validate_return_type(template_value, return_value)
+                                return return_value
 
                             callable_value = awaitable_return_validation_wrapper
```

Comment:
Have you tried wrapping `callable_value` for both cases here, and inside the wrapper, do the return type validation?

This is how it's done at mock_callable: https://github.com/facebookincubator/TestSlide/blob/22ed7af503cb9b42caa9a12ff8d8f3e88934da01/testslide/mock_callable.py#L473-L486.

Reply:
![](https://media.giphy.com/media/wzD3nQPA4gqHK/giphy.gif)